### DISCUSSION TITLE
Circular dependency of stores not working

### DISCUSSION THREAD

**Discussion by Thomas-1985** [vuejs/pinia#2229] [Help and Questions]
We want to migrate from Vuex 3.5 to Pinia and have a problem with stores, which call each other.
We have a `rootStore` and an `newStore` in Pinia.

In an action of the `rootStore`, we try to use `const newStore = useNewStore();`
Otherwise in an action of the the new store, we try to use `const rootStore = useRootStore();`

This will not work, in the root store we get the error `TS2448: Block-scoped variable 'useNewStore' used before its declaration.`
I guess this is because in the `defineStore` calls, each other function is called, creating a circular dependency.

The pinia introduction says: "You can even have circular dependencies of stores"

Also we have read that this is possible here: https://pinia.vuejs.org/cookbook/composing-stores.html
But the example is using a `setup()` function for the stores, we use the same option structure as vuex (with data, getters, actions and so on) with the options api (with Vue 2.7)

How can we solve this? Thanks for your help.

**Comment by guyaumetremblay** [comment]
@Thomas-1985 Did you solve this? I have the same scenario but not the same error. Mine is : `ERROR in Circular dependency detected` ! The compiler doesn't like the circular dependency of the imports...

**Reply by pavitra-infocusp** [reply]
Try using lazy imports

**Comment by renatodeleao** [comment]
> [!NOTE]
> EDIT: 🤦 i've realized that the issue is not with pinia, but just regular javascript circular references, you can't obvious mutually import modules _this way_. Solving circular deps issues in big apps is a non-trivial problem and there's plenty of articles around. I've found a workaround by using a pinia plugin that exposes a `$stores` property via the pinia instance private property that holds the map, but I don't advise doing this as it can break on version upgrades. Probably better to architecture your stores differently. Hopefully some smarter folks will post here their approaches.

Have the same issue! Here's the fork of the [the pinia todo app in the pinia-playground](https://play.pinia.vuejs.org/#eNqtV9tuGzcQ/ZWJXiTH8m4Mp0EhSIbTNAWch7RInKcoKFZarsV4l1yQXNmGK6Af0cd+QL8rX9LDy15tC0abF0kkZ84M53KGuhu9LstoW7HRbDTXa8VLQ5qZqjxdCl6UUhm6I8Uy2lGmZEFjiI47R9pIxS7kB5bpRqTkgiddoUpDJJWtRBQbu46+6oHYJ81URwo7ygstxVoKbcjpfbRGadHgTg7qY6vQObVw7eEdZTw3TE3DN0trrxbde0xaG1CtlQW7ttIX7MZAHBGZjMfuOKvE2nApKEkd3uSA7paCiGc0edbRirZJXrFwSEAwlYI60c6iUOdmUY10X9sK3tuFP2OEyOFYZ2XOolxeTppgRCIprO489hlGbrEwrCjzxDCsiOab49M7l4IQv91uHmPPneXJiuWncy7KytD2qJApyxfLkY/ickTmtmRYqyTlEkvnFNZJni9Hp/Q6z+exh/g/aBmqSm9YaiF/Cb+/B24lusifmlUPe6P8d+XX1hqHjUwqINjMERf9qoKF2RW7DccRt+heFcpDH72Irq/VOrvesPXVSt50lJElJ25sJe6QdIcY59x7GAcXQzj88Xt2TdarWZAe2u+UVGvbGsDqDLcoFdM6YgLXs3n15Rl86sVpVRmDVpilXCernKWQ7vaAhVvnfH3VQ3mdItheEyjzuCnM0XTkyeGoSErwgBRgKdc/y3CAMM/qjlqOHPHYneUodr8jpoujlZLXnkeWo2ktegYei1O2NVLm+igpuVfbGFPqWRyvUwF5BIdvVSSYiUVZxPd0zl5FP0THCP0qhqGYi5Td9M1A4yhlxVPQa9GzF9Hxy+iVQ92eBNBiiPpUd1UlDC+ALIuzk+hldPJjjNyY7v7jQbIA2NwydaQY/FC2j55md6DWsz04u2c/kOIO6TcahJbxy0Hy17Ioec7Ur6Vl3n4RgHXk9Tu3Z1TFmsu4Vnpg/6tGd9lL/YYit551AmASdclQtfb47cf3roKbQzRPlYdEPHL4gYGMK+ujF/upEinc7sg5b89dLXNxeaHf3hgmdH0p66iLhpN3CXmz5+qtuyfRSTeKYdgigM2wTRmIjgWu3zO490xkduPEmsnrp+miCz0ZO9vjqXdSG7T1jDAiF6c0CX7Hz5/TmeUcR20I4wxyCtGYEk9nJKpiZWd2S48zWqEDWSKmbl793krR7vOXHT2PPbAzPaPPX0K0O4bGiPuY/qBxzfZu0Y6CcYviWX1GTqVGih1J0jXPc1oxSioji8TwNWRuMQoypjAH4EDwy2shYOYcvr5wKLsD94WEAR5uhmjUHviHjQtY82Zwhq0tW/85swXwjL79/U996t8UPsqRT7r3fuLeNC7qg1HjXhR1gRG1EXjY/lMsPNtvAlkIaMiGx9PfKfM1bi9zj13Fvs/MhtcXoMUCz6gm/x1BF3WbWZJmA8GQMSTfbAbZ6CajCVYw0glrLbIjlmv2sCedWuz50gUdZKuBHWTLybpkdTPhPxP3dm3LzxaYuKWkkBgQJDMCpVQFJr+e1nAJlSADDselIiGNV6zfrDaNHYf3Psub6N7KChEWlHLF1gYtVFQ2W/Cc+VqrZdurRGWlN5M7ak5c+bhH0SGN6duff+HzsLXs3sC+rhyIb8bDw355ZQkS0imuVp2n6Nhh/HYHYNdAif+FXPf8K7pPr46JB+xqbe8lV3uJ49B5NgLgsXdyI+hnaf/GPcpCxa1z7Y2tgodIiIMpMLxxIkBzm8TQORyt8pTwFLbxJTSDzZ67EflpOCU88ghvZqRJ23FHZaVKqVlTuRY589UgC1t1VKDKoGTN1NwjCwYLliSsgaQsPQ1fS3XVr7pH/yv2u6P939UjNPeqrwkzFIRrTU9+PD2IciYuzSYURT1vD0a7fwG8jykW)

- Both `user` and `todos` store call each other.
- I've tried using a barrel file for the stores imports, as well but the result is the same.

- I've tried same demo using [setup stores syntax](https://play.pinia.vuejs.org/#eNqtV9tuGzcQ/ZWJXiTb8m4Mp0EhSK7TNAWSh7TI5SkKipWWshjvkguSK9twBfQj+tgP6HflS3p42V1qXTlGkgdfSM6cOZwZzszeDp5VVbKp2WAymOql4pUhzUxdnc0FLyupDN2SYiva0krJkoYQHUZH2kjF3sk3bKVbkYoLnsVCtYZILjuJJDV2nXzSPbH3mqlICjvKC83FUgptyOm9tUZp1uKODppjqxCdWrju8JZWvDBMjcNfljesZvE9Rp0NqDbKgl1Z6Xfs2kAcHhkNh+54VYul4VJQlju80QHdzgURX9HoUaSVbLKiZuGQgGBqBXWirUWh6GZJg3RX2wre2QWfIVzkcCxZWbCkkBej1hmJyEqrO019hBFbLAwrqyIzDCui6frk7NaFIFKh7Xaa4sAJFNmCFWdTLqra0Oa4lDkrZvOBd+V8QOamYlirLOcSS8cM66wo5oMzelYU09RDfAvaCqml1yy3kL+G/78Hbi1i5Pftagd7rfzf2q+tNQ4bK6mAYMNHXOymFixMLtlNOE64RfeqUO5z9CK6uVZHdrlmy8uFvI6UESonbmw6bhF5h5gW3DNMA8XgDn/8ml2RZTUJ0n37UV51tq0BrM5xi0oxrRMmcD0bV5+jgdOOnxa1MXgPk5zrbFGwHNLxQ7Bwy4IvL3dQnuVwttcEyjRts3MwHvgKcVxmFYqBFChV7hHNwwHcPGme1Xzgqo/dmQ9S93/CdHm8UPLKF5P5YNyInqOYpTnbGCkLfZxV3Kutjan0JE2XuYA8nMM3KhHMpKIq0zs650+TH5ITuH6RwlDKRc6ud81A4zhn5UPQG9Hzx8nJk+SpQ92cBtCyj/pQuqoWhpdAluX5afIkOf0xRWxMvL/fSRYAmxumjhUDD2Xf0cPs9tR2bPfO7tgPlXGL8BuNqrbiF73gL2VZ8YKp3ypbfneTAFVHXr1ye0bVrL2Me0r/s/9J43XZS/2OJLfMIgeYTF0wZK09fvH2tcvg9hCPpy5CIPYcvmGoyLXl6MV+rkUO2pGcY/vS5TIXF+/0i2vDhG4uZYk6bzh5F5Dn91y9o3uanMZeDB0XDmw7bs5Q6JhvmPd1771tORoRMnTBDRuTDUttWD5GhdoZC/zkMBfs2um0Dds34VlMZjR0bIdjQjednfnL3dvhOwFtUDVci/aMRsEz6eEhnduq5oonAjWBqIK/x8TzCYm6XNjRoCvAE1rgjbNMjJ3RPzop2n74uKXD1AM7qhP68DHEMzI0RGSH9CcNm37iFl2zGXYovm9MyKk0SKkrw3TFi4IWjLLayDIzfAmZGzSbFVPoNCAQeHktONi8BNfHDmXrZgb7AzQkBozozlkNkSYGTfBGkd+dpjVtDwtmM+4Rff7n33iO8V5PfI75q4zcHOVgep3NMbK8oqi2LrmfyUOsPdpvzrnh8NBjIUoeTn+njAiwLqCNe+M5c9+l7KTob+TlaTbDRNfmSDsyukDY2JM0a4iFaCI9zLoXoC4+rc92HOwHRy+xJVZotodFlKsRjwDZi1qA2xepO9lo3yfcFGqX3RA3lJUSXYnkilDH6hLjhh43YBlVqCQcZKUiIY1VuzN+20C2VAF6I2s4TVDOFVsavJuydhUCLvTswitecx0yqqr1enTrWGLf5YWbs45oSJ//+hu/j3qDsk8YB+Ff39HRbt6sMrg4yppOnefBKeEzIFw18E8ScHIfJY7qQSgMO24Pe71gtJJRBoa9ZugKveEAzSFU9K/pDQ/5svuG7uBaT685WLa7vQFx5vAARgn4SaAkrjNDL4FSFzlhMLehISSJDbojSL43jwkjJ2GCR4S1bb5U1aqSmrlstagrn0CytKlI+CaSULAmmgSXJQO6LRoWPKsqX66vpLoMGF/4cv1y67IZcxKCZ9MNXeKVXAv6RVqfxcXN45Q3Dvy5e0lfKKUdrZ92C6r7oGmKd0hcVxP8k+b5QVIwcWHWcTV/UPbG9Lok3P4Hyd+g6w==)

Anyone figured it out?

Cheers ✌️

**Reply by renatodeleao** [reply]
Follow-up.

I'm migrating a very large app from `vuex` to `pinia` that does have a circular references between two modules (now stores). In vuex, we can access other modules via `rootState/rootGetters` exposed in `getters/actions` as function arguments since each module is merged into a global single store, something we don't have _publicly_ in pinia AFAIK — we do have a private `Map` of the registered stores in the `pinia` instance, but it's undocumented.

**it's important to note** that, as always in programming, we can do things differently, but the point here is that docs say that [_we can have circular references between stores_](https://pinia.vuejs.org/introduction.html#Comparison-with-Vuex-3-x-4-x) but the docs demo-ed way is if the two stores are in the same file which is a bit impractical in a real world project. 

### Demo
[DEMO APP](https://stackblitz.com/edit/vitejs-vite-gajaue?file=src%2FApp.vue). 

The demo it's slightly more advanced "todos" app, a `todos` store  with a `users` store that cross-reference each other. In this app I introduce the concept of actions side-effects in the stores (the ones that required cross-access). This kind of side-effects are very common in ORMs that work with normalised state (like `pinia-orm`), but for the sake of demo simplicity we did it manually.

1. **Action:** assigned a todo to a specific user. 
   **Side-effect:** needs to see if user is "allowed to take that task" before actually doing it (the demo limits 3 tasks per user)
1. **Action:** Destroying a user. 
   **Side-effect:** un-assign all user’s tasks before removing it.



The way i've workaround it for the migration sake is via plugin, dangerously accessing the registered stores `Map` in the pinia instance private property. The other major downside of this approach is that you need to call explicitly call `useUsersStore()` even if you just wanted to call `useTodosStore` in template, because that's how pinia registers the store in the singleton, but that can be workaround again with a plugin.

Hope it helps someone! 
Also hope that a better programmer finds this and propose a better way. 

My other alternative way was dependency injection `useStoreA(useStoreB)`, `useStoreB(useStoreA)`. W

```js
//storeA.js pseudo-code

export function useStoreA = (useStoreB) => {
   return defineStore('a', () => {
     // access dependency store here
     const storeB = useStoreB()
   })()
})
```

Cheers!

**Reply by CyberAP** [reply]
@renatodeleao I solved it with a plugin. Do not use this as a regular pattern.

```js
// plugin.js
export const globalAccessorPlugin = (context) => {
  context.store.tryStore = (storeName) => {
    const anotherStore = context.pinia._s.get(storeName);

    if (!anotherStore) {
      const hookName = `use${storeName.charAt(0).toUpperCase() + storeName.slice(1)}`;
      throw new ReferenceError(
        `Store '${storeName}' doesn't exist. Ensure you have called ${hookName}.`,
      );
    }

    return anotherStore;
  };
};
```

```js
// store.js
import { defineStore } from 'pinia';

export const useMyStore = defineStore('myStore', () => {
  getters: {
    derived() {
      return this.tryStore('anotherStore').anotherState + 1;
    }
  }
});
```

Don't forget to register it:

```js
createPinia().use(globalAccessorPlugin)
```

**Reply by renatodeleao** [reply]
Thanks for the reply @CyberAP. Yeah that's how I end up solving this too — for the [demo i did very barebones one with zero safeguards](https://stackblitz.com/edit/vitejs-vite-gajaue?file=src%2Fmain.js).

I've tried posting on the `#pinia` vue discord channel to see if I got "safer" approaches from the community but got a grand total of 0 😅 

I feel like these mutual ORM-like side-effects between stores is pretty common/standard in apps that use optimistic UIs, so I would like to see some guidance in docs on the _"[we have circular references between stores](https://pinia.vuejs.org/introduction.html#Comparison-with-Vuex-3-x-4-x)"_ matter.

Also tried to dig into the `pinia-orm` codebase to see how they've achieved it, but couldn't fully grasp it. I think they register all stores into the `entities` store and use that to proxy access it becomes the root store with access to all others.
